### PR TITLE
Add support for Java modules in Eclipse dependencies

### PR DESCRIPTION
**Pull request by jjohannes**
The Java module detection service is used to set the 'module=true' attribute on Eclipse classpath entries that represent Java modules. This will give use a similar behavior to Gradle's compile task
in Eclipse.

### CODE REVIEW

**Comment by jjohannes** [comment]
@bot-gradle test QF

**Comment by jjohannes** [comment]
@bot-gradle test QF

**Comment by jjohannes** [comment]
@bot-gradle test QF

**Comment by jjohannes** [comment]
@bot-gradle test QF

**Comment by jjohannes** [comment]
@bot-gradle test QF

**Comment by jjohannes** [comment]
@bot-gradle test this

**Comment by jjohannes** [comment]
@bot-gradle test RFM

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java [hunk 1]
```diff
@@ -240,6 +241,7 @@ private static void configureJavaClasspath(final Project project, final TaskProv
             @Override
             public void execute(JavaPlugin javaPlugin) {
                 model.getClasspath().setPlusConfigurations(Lists.newArrayList(project.getConfigurations().getByName("compileClasspath"), project.getConfigurations().getByName("runtimeClasspath"), project.getConfigurations().getByName("testCompileClasspath"), project.getConfigurations().getByName("testRuntimeClasspath")));
+                model.getClasspath().getModularClasspathHandling().getInferModulePath().convention(project.getExtensions().getByType(JavaPluginExtension.class).getModularClasspathHandling().getInferModulePath());
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:244]: nit: unnecessary whitespace

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java [hunk 2]
```diff
@@ -145,9 +148,12 @@
 
     private final org.gradle.api.Project project;
 
+    private final ModularClasspathHandling modularClasspathHandling;
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:151]: ❌ Why do we need this here? As far as I can tell there should not be any Eclipse-specific configuration with regards to the module system. I'd try to remove this from the `eclipse` plugin's public API.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java [hunk 3]
```diff
@@ -214,24 +219,36 @@ private void resolveAuxiliaryArtifacts(IdeDependencyVisitor visitor) {
...
+                        visitor.visitGradleApiDependency(artifact, localGroovySources, testOnly);
+                    } else {
+                        visitor.visitFileDependency(artifact, testOnly);
+                    }
                 }
             }
         }
 
+        private boolean isModule(boolean testOnly, File artifact) {
+            // Test code is not treated as modules, as Eclipse does not support compiling two modules in one project anyway.
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java:244]: 👍

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java (hunk 2 above)
Comment (jjohannes) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java]: You are right. I did this in the beginning when I didn't have the full picture and then I did not question it again. I removed the API and now look up the value in the Java compile task the moment we resolve the classpath for Eclipse.

It's also a bit more correct this way. If we look at the task, we will also get the right value if it is only set for the task (and not on the extension).

**Comment by jjohannes** [comment]
@bot-gradle test this
